File: run-migration.py
#!/usr/bin/env python3
"""
Migration Script: Add Revolico Import Fields
Adds source, revolicoId, scrapedAt fields and makes sellerId nullable
"""

import os
import sys

# Read DATABASE_URL from environment or use default
database_url = os.getenv('DATABASE_URL', 'postgresql://postgres:password@localhost:5432/cuba_marketplace')

print('Starting migration for Revolico import fields...')
print(f'Database: {database_url.split("@")[-1] if "@" in database_url else "local"}')

# Read SQL file
with open('migrate.sql', 'r') as f:
    sql = f.read()

# Use psycopg2 if available, otherwise use subprocess
try:
    import psycopg2

    conn = psycopg2.connect(database_url)
    cur = conn.cursor()

    print('\n1. Making sellerId nullable...')
    print('2. Adding source field...')
    print('3. Adding revolico_id field...')
    print('4. Adding scraped_at field...')
    print('5. Creating index on revolico_id...')
    print('6. Creating index on source...')

    cur.execute(sql)
    conn.commit()

    print('\n✅ Migration completed successfully!')
    print('Added fields:')
    print('  - source (TEXT, default: "user")')
    print('  - revolico_id (TEXT)')
    print('  - scraped_at (TIMESTAMP)')
    print('Modified fields:')
    print('  - seller_id (now nullable)')
    print('Created indexes:')
    print('  - idx_listings_revolico_id')
    print('  - idx_listings_source')

    cur.close()
    conn.close()

except ImportError:
    print('psycopg2 not found, using psql command...')
    import subprocess

    # Extract connection details from DATABASE_URL
    try:
        from urllib.parse import urlparse
        result = urlparse(database_url)

        env = os.environ.copy()
        if result.password:
            env['PGPASSWORD'] = result.password

        cmd = [
            'psql',
            '-h', result.hostname or 'localhost',
            '-p', str(result.port or 5432),
            '-U', result.username or 'postgres',
            '-d', result.path[1:] if result.path else 'cuba_marketplace',
            '-f', 'migrate.sql'
        ]

        subprocess.run(cmd, env=env, check=True)

        print('\n✅ Migration completed successfully!')

    except Exception as e:
        print(f'❌ Migration failed: {e}')
        sys.exit(1)

except Exception as e:
    print(f'❌ Migration failed: {e}')
    sys.exit(1)

File: run-migration.ts
/**
 * Migration Script: Add Revolico Import Fields
 * Adds source, revolicoId, scrapedAt fields and makes sellerId nullable
 */

import { Pool } from 'pg';
import * as dotenv from 'dotenv';

dotenv.config();

const pool = new Pool({
  connectionString: process.env.DATABASE_URL,
});

async function runMigration() {
  const client = await pool.connect();

  try {
    console.log('Starting migration for Revolico import fields...');

    // Start transaction
    await client.query('BEGIN');

    // Make sellerId nullable (drop NOT NULL constraint)
    console.log('1. Making sellerId nullable...');
    await client.query(`
      ALTER TABLE listings
      ALTER COLUMN seller_id DROP NOT NULL;
    `);

    // Add source field
    console.log('2. Adding source field...');
    await client.query(`
      ALTER TABLE listings
      ADD COLUMN IF NOT EXISTS source TEXT DEFAULT 'user';
    `);

    // Add revolicoId field
    console.log('3. Adding revolico_id field...');
    await client.query(`
      ALTER TABLE listings
      ADD COLUMN IF NOT EXISTS revolico_id TEXT;
    `);

    // Add scrapedAt field
    console.log('4. Adding scraped_at field...');
    await client.query(`
      ALTER TABLE listings
      ADD COLUMN IF NOT EXISTS scraped_at TIMESTAMP;
    `);

    // Create index on revolico_id for faster lookups
    console.log('5. Creating index on revolico_id...');
    await client.query(`
      CREATE INDEX IF NOT EXISTS idx_listings_revolico_id
      ON listings(revolico_id);
    `);

    // Create index on source for faster filtering
    console.log('6. Creating index on source...');
    await client.query(`
      CREATE INDEX IF NOT EXISTS idx_listings_source
      ON listings(source);
    `);

    // Commit transaction
    await client.query('COMMIT');

    console.log('\n✅ Migration completed successfully!');
    console.log('Added fields:');
    console.log('  - source (TEXT, default: "user")');
    console.log('  - revolico_id (TEXT)');
    console.log('  - scraped_at (TIMESTAMP)');
    console.log('Modified fields:');
    console.log('  - seller_id (now nullable)');
    console.log('Created indexes:');
    console.log('  - idx_listings_revolico_id');
    console.log('  - idx_listings_source');

  } catch (error) {
    // Rollback on error
    await client.query('ROLLBACK');
    console.error('❌ Migration failed:', error);
    throw error;
  } finally {
    client.release();
    await pool.end();
  }
}

runMigration().catch((error) => {
  console.error('Fatal error:', error);
  process.exit(1);
});
